cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i

```diff
@@ -286,32 +286,77 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute
+    if (valueNT)
+    {
+        values.reset(nDataBlocks);
+        DAAL_CHECK_MALLOC(values.get());
+    }
+
+    TlsMem<algorithmFPType, cpu> tlsLogP(nRowsInBlock * nClasses);
+    TArrayScalable<algorithmFPType, cpu> grads;
+    if (gradientNT)
+    {
+        grads.reset(nDataBlocks * nBeta);
+        DAAL_CHECK_MALLOC(grads.get());
+    }
+    const algorithmFPType div = algorithmFPType(1) / algorithmFPType(n);
+    const bool bL1 = (parameter->penaltyL1 > 0);
+    const bool bL2 = (parameter->penaltyL2 > 0);
     if (valueNT || gradientNT || hessianNT)
     {
-        //f = X*b + b0
-        applyBetaThreaded(x, b, f.get(), n, nClasses, p, parameter->interceptFlag);
+        daal::threader_for(nDataBlocks, nDataBlocks, [&](size_t iBlock) {
+            const size_t iStartRow      = iBlock * nRowsInBlock;
+            const size_t nRowsToProcess = (iBlock == nDataBlocks - 1) ? n - iBlock * nRowsInBlock : nRowsInBlock;
 
-        //f = softmax(f)
-        softmaxThreaded(f.get(), f.get(), n, nClasses);
+            const algorithmFPType* xLocal = x + iStartRow * p;
+            const algorithmFPType* yLocal = y + iStartRow;
 
-        const bool bL1 = (parameter->penaltyL1 > 0);
-        const bool bL2 = (parameter->penaltyL2 > 0);
+            algorithmFPType* fPtrLocal    = f.get() + iStartRow * nClasses;  // potential opt 1
 
-        const algorithmFPType div = algorithmFPType(1) / algorithmFPType(n);
-        if (valueNT)
-        {
-            TArrayScalable<algorithmFPType, cpu> logP(f.size());
-            daal::internal::Math<algorithmFPType, cpu>::vLog(n * nClasses, f.get(), logP.get());
+            //f = X*b + b0
+            applyBeta(xLocal, b, fPtrLocal, nRowsToProcess, nClasses, p, parameter->interceptFlag);
+
+            //f = softmax(f)
+            softmax(fPtrLocal, fPtrLocal, nRowsToProcess, nClasses);
+
+            if (valueNT)
+            {
+                algorithmFPType* logP = tlsLogP.local();
+                daal::internal::Math<algorithmFPType, cpu>::vLog(nRowsToProcess * nClasses, fPtrLocal, logP);
+
+                algorithmFPType localValue(0);
+                for (size_t i = 0; i < nRowsToProcess; ++i)
+                {
+                    for (size_t j = 0; j < nClasses; ++j) localValue += (size_t(yLocal[i]) == j) * logP[i * nClasses + j];  // potential opt 2
+                }
+                values[iBlock] = localValue;
+            }
 
+            if (gradientNT)
+            {
+                DAAL_ITTNOTIFY_SCOPED_TASK(applyGradient);
+                algorithmFPType * g = grads.get() + iBlock * nBeta;
+                const algorithmFPType interceptFactor = (parameter->interceptFlag ? 1 : 0);
+                for (size_t i = 0; i < nBeta; ++i) g[i] = 0;
+                for (size_t i = 0; i < nRowsToProcess; ++i)
+                {
+                    addGradInPt<algorithmFPType, cpu>(g, xLocal + i * p, fPtrLocal + i * nClasses, size_t(yLocal[i]), interceptFactor, nClasses, nBetaPerClass);
```

Comment:
Can we replace this with GEMM?

CC: @PetrovKP

Reply:
If allocate matrix K = { nRowsToProcess x nClass } and fill this by this rule ```((y[i] == k) ? (pi[iC] - algorithmFPType(1.)) : pi[iC]);```
then we can use GEMM Kt x X to fill gradient matrix for all block

Reply:
As I understand it's already implemented in oneAPI code

Reply:
Yes, see [here](https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/objective_function/cross_entropy_loss/oneapi/cross_entropy_loss_dense_default_oneapi_impl.i#L350)